light/provider/errors.go

```diff
@@ -41,3 +41,11 @@ type ErrUnreliableProvider struct {
 func (e ErrUnreliableProvider) Error() string {
 	return fmt.Sprintf("client deemed unreliable: %s", e.Reason)
 }
+
+// ShouldBeRemoved analyzes the nature of the error and whether the provider
+// should be removed from the light clients set
+func ShouldBeRemoved(err error) bool {
```

Comment:
nit: can we call this 'IsBadWitness'? There might be many different things we'd want to do with a bad witness.

Reply:
Maybe it's more appropriate to move this function to the light package than in the provider package but I think it's helpful to call this by what it does to the provider. Just calling it `IsBadWitness` seems rather vague (aren't all errors bad anyway). I also can't imagine that there would be different things we'd want to do. Either we remain connected or not.

Reply:
A reason for my comment was that the error that we create as a result of this check is `errBadWitness`. But I would agree, it seems that this is relating to the logic of light client in a way that the `provider` code does not need to have any understanding of. Perhaps a more reasonable thing to do would be have the caller of `compareNewHeaderWithWitness` handle removing the peer and therefore call this error method. Otherwise the method is making assumptions about its caller's behavior which is not ideal.

looking at the `compareNewHeaderWithWitness` call sites, it seems that they are in a go routine and don't actually check the error though, so I'm a bit confused what this is doing at all?

Reply:
The go routines pass the error through a channel to the main thread which then processes them sequentially. 

I will move this function to the light client package

Reply:
Missed that! sounds good.